src/error.rs

```diff
@@ -8,6 +11,15 @@ pub enum DaftError {
     ComputeError(String),
     ArrowError(String),
     ValueError(String),
+    // ExprResolveTypeError: Typing error when resolving expressions against schemas
+    //
+    // This variant has custom Display logic for presenting a more user-friendly error message which shows
+    // exactly which operation, arguments and dtypes of those arguments caused the issue.
+    ExprResolveTypeError {
+        expectation: String,
+        expr_repr: String,
```

Comment:
any reason this isn't a ExprRef and format it in the display?

Reply:
I notice everywhere its just `self.to_string()` might as well format it fmt::Display here and skip the intermediate String

Reply:
It's a little tricky to get this to work with `AggExpr` if I do that, can show you offline but here's the location where it's tricky to get it to work: https://github.com/Eventual-Inc/Daft/pull/690/files#diff-52451b0b516f3fc9684de696a3b804b4227c83092c31fcaf2653b6dbf525619dR72-R77

Because we raise the error from AggExpr, we actually don't have a handle to the parent `ExprRef` to pass it in.

Reply:
that should be fine I think you should be able to do 
`Arc::new(Expr::Agg(self))`

Reply:
or if you want to give it ownership you can do just an Expr in the error. can give it the Expr to own

Reply:
I guess this is going to be a problem too with functional expressions. How do you plan to handle that right now?

Reply:
I was planning on having functional expressions pass in `.to_string()` as well.

But I think your suggestion of wrapping itself in an `Arc::new(Expr::FunctionExpr/Agg(self))` might make sense too, I didn't think of that. Let me try it out